hazelcast/src/main/java/com/hazelcast/map/impl/operation/AddIndexOperation.java

```diff
@@ -91,6 +91,8 @@ public void runInternal() {
 
         SerializationService serializationService = getNodeEngine().getSerializationService();
 
+        index.beginPartitionUpdate();
```

Comment:
When an index is being built, we index partitions one by one. The query fragment that should be executed on the given member, has a set of expected partitions, say `[1, 2, 3, 4]`. Unless all of them are indexed, the SQL query will report an error. This 'beginPartitionUpdate' adds one more check: if the partitions `[1, 2, 3, 4]` are already indexed *AND* another partition is being indexed at the moment, we disallow querying indexes as well, because it may return more entries than needed.

Does it answer your question?

Reply:
Sure, I was wondering generally in the context of the old query engine though, but if you don't know, that's ok.

Reply:
I think I didn't understand your question in my first answer.
If your question is "how come that the old predicate API worked", then I think that it is subject to the same inconsistency issue, but it is harder to reproduce because the old predicate API does not hold index iterators opened for the prolonged time. As mentioned in other comments, I'll create a follow-up ticket for the predicate API fix.